serverless/client/witness/consensus.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package witness provides client support for using witnesses.
+package witness
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/formats/log"
+	"github.com/google/trillian-examples/serverless/client"
+	"golang.org/x/mod/sumdb/note"
+)
+
+// CheckpointNConsensus returns a ConsensusCheckpoint function which selects the newest checkpoint available from
+// the available distributors which has signatures from at least N of the provided witnesses.
+func CheckpointNConsensus(logID string, distributors []client.Fetcher, witnesses []note.Verifier, N int) client.ConsensusCheckpoint {
+	// TODO(al): This implementation is pretty basic, and could be made better.
+	// In particular, there are cases where it will fail to build consensus
+	// where a more thorough algorithm would succeed.
+	// e.g. currently, it will simply fetch  checkpoint.N from each of the
+	// distributors and fail if either:
+	//  - no distributor has a checkpoint.N file
+	//  - no distributor has a checkpoint.N file signed by N of the known witnesses.
+	//
+	// It's good enough for now, though.
+	return func(ctx context.Context, logSigV note.Verifier, origin string) (*log.Checkpoint, []byte, error) {
+		type cp struct {
+			cp  *log.Checkpoint
+			n   *note.Note
+			raw []byte
+		}
+		cpc := make(chan cp, len(distributors))
+		wg := &sync.WaitGroup{}
+		for _, f := range distributors {
+			wg.Add(1)
+			go func(ctx context.Context, f client.Fetcher, logID string, N int, cpc chan<- cp) {
+				defer wg.Done()
+				c, n, cpRaw, err := getCheckpointN(ctx, f, logID, N, logSigV, origin, witnesses)
+				if err != nil {
+					glog.Infof("Error talking to distributor: %v", err)
+					return
+				}
+				cpc <- cp{
+					cp:  c,
+					n:   n,
+					raw: cpRaw,
+				}
+
+			}(ctx, f, logID, N, cpc)
+		}
+
+		wg.Wait()
+		close(cpc)
+
+		glog.Infof("considering %d cps", len(cpc))
+
+		var bestCP cp
+		for cp := range cpc {
```

Comment:
This is a bit ugly that you're using `cp` to be a local variable when it was defined as a type up there and a variable of that type created just the line above this. Maybe just `c`?

Reply:
done